src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -15,14 +15,70 @@ private BoundExpression BindIsPatternExpression(IsPatternExpressionSyntax node,
         {
             var expression = BindExpression(node.Expression, diagnostics);
             var hasErrors = node.HasErrors || IsOperandErrors(node, expression, diagnostics);
-            var pattern = BindPattern(node.Pattern, expression, expression.Type, hasErrors, diagnostics);
+            var pattern = BindPattern(node.Pattern, expression, ExpressionIsNull(expression), expression.Type, hasErrors, diagnostics);
             return new BoundIsPatternExpression(
                 node, expression, pattern, GetSpecialType(SpecialType.System_Boolean, diagnostics, node), hasErrors);
         }
 
+        /// <summary>
+        /// Returns true if the expression is known to always be null.
+        /// Returns false if the expression is always known to be non-null.
+        /// Returns null otherwise.
+        /// </summary>
+        /// <param name="expression"></param>
+        /// <returns></returns>
+        internal static bool? ExpressionIsNull(BoundExpression expression)
+        {
+            if (expression.ConstantValue != null)
+            {
+                return expression.ConstantValue.IsNull;
+            }
+
+            if (expression.Type == null || expression.Type.IsNonNullableValueType())
+            {
+                return false;
+            }
+
+            switch (expression.Kind)
+            {
+                case BoundKind.Conversion:
+                    {
+                        var conversion = (BoundConversion)expression;
+                        switch(conversion.ConversionKind)
+                        {
+                            case ConversionKind.AnonymousFunction:
+                            case ConversionKind.Boxing:
+                            case ConversionKind.ImplicitNullable:
+                            case ConversionKind.InterpolatedString:
+                            case ConversionKind.MethodGroup:
+                                return false;
+                            case ConversionKind.ExplicitReference:
+                            case ConversionKind.Identity:
+                            case ConversionKind.ImplicitReference:
+                                return ExpressionIsNull(conversion.Operand);
+                            case ConversionKind.NullLiteral:
+                                return true;
+                            default:
+                                return null;
+                        }
+                    }
+                case BoundKind.AnonymousObjectCreationExpression:
+                case BoundKind.ArrayCreation:
```

Comment:
I wonder how precise do we need to be here with "can it be null". Are we only applying constant propagation rules as in C# spec, use common sense or platform rules?

For example "this" and "new ReferenceType()" are never null from the language prospective, but underlying platform allows for edge cases where they can be null. 

If this is not just an optimization, but there is a semantical impact here, we would need to document all these rules.